### PR TITLE
compliance_more_filtering_criteria_for_consolidated_sanctions_list

### PR DESCRIPTION
**Pull request by mat-fs**

### CODE REVIEW

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 1]
```diff
@@ -126,28 +172,35 @@ sub _process_name_and_dob {
...
 
-    for my $name (@$name_list) {
+    delete $data{date_of_birth};
+    $data{dob_epoch} = \@dob_epoch;
+    $data{dob_year}  = \@dob_year;
+    $data{dob_text}  = \@dob_text;
+
+    # convert all country names to iso codes
+    for my $field (qw/place_of_birth residence nationality citizen/) {
+        $data{$field} = [map { my $val = trim $_; lc(code2country($val) ? $_ : country2code($val) // '') } $data{$field}->@*];
```
Comment (felipe-binary) [inline lib/Data/Validate/Sanctions/Fetcher.pm:191]: If `code2country($val)` evaluates to false (`undef` I guess), you don't to call it again in the `else` of the ternary operator?

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 2]
```diff
@@ -104,15 +106,59 @@ sub _date_to_epoch {
...
-    return eval { Date::Utility->new($date)->epoch; };
+    my $result = eval { Date::Utility->new($date)->epoch; };
+    return $result;
 }
 
-sub _process_name_and_dob {
-    my ($name_list, $dob_list, $dataset) = @_;
+=head2 run
+
+Proocesses an entry retrived from sanction resouces and saves it into the specified key-value dataset.
```
Comment (felipe-binary) [inline lib/Data/Validate/Sanctions/Fetcher.pm:115]:
```suggestion
Processes an entry retrieved from sanction resources and saves it into the specified key-value dataset.
```

File: lib/Data/Validate/Sanctions/Fetcher.pm [hunk 3]
```diff
@@ -199,35 +283,51 @@ sub _hmt_csv {
...
+        $parsed = $csv->parse($line);
+        next unless $parsed;
 
         my @row = $csv->fields();
-        my $row = \@row;
-        ($row->[23] and $row->[23] eq "Individual") or next;
-        my $name = _process_name @{$row}[0 .. 5];
 
-        next if $name =~ /^\s*$/;
+        @row = map { trim($_ =~ s/\([^(]*\)$//r) } @row;
```
Comment (felipe-binary) [inline lib/Data/Validate/Sanctions/Fetcher.pm:306]: does this erases everything enclosed between parentheses (included) at the end of each row? (wild guess)

File: lib/Data/Validate/Sanctions.pm [hunk 4]
```diff
@@ -41,7 +44,7 @@ sub update_data {
 
     my $updated;
     foreach my $k (keys %$new_data) {
-        if (ref($self->{_data}{$k}) ne 'HASH' || $self->{_data}{$k}{updated} < $new_data->{$k}{updated}) {
+        if (!$self->{_data}{$k} || $self->{_data}{$k} ne 'HASH' || Compare($self->{_data}{$k}, $new_data->{$k})) {
```
Comment (chylli-binary) [inline lib/Data/Validate/Sanctions.pm:47]: if `$self->{_data}{$k}` , `$new_data->{$k}` is same, then we update it ?
I guess this logic should be reversed.

File: lib/Data/Validate/Sanctions.pm [hunk 5]
```diff
@@ -112,37 +211,37 @@ sub get_sanctioned_info {    ## no critic (RequireArgUnpacking)
         my @names = keys %{$data->{$file}->{names_list}};
 
         foreach my $sanctioned_name (sort @names) {
+            my $entry = $data->{$file}->{names_list}->{$sanctioned_name};
```
Comment (kirill-binary) [inline lib/Data/Validate/Sanctions.pm:214]: Is it possible to have multiple entries in the same sanction list with the same name? if so we should not merge them.

File: lib/Data/Validate/Sanctions/Fetcher.pm (hunk 1 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions/Fetcher.pm:191]: It's not called twice. The second one is just the inverse: `country2code`. Country can be a code or name, depending on the data source. I'm converting all of them to codes here.

File: lib/Data/Validate/Sanctions.pm (hunk 4 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm:47]: You are right. I'll correct it.

File: lib/Data/Validate/Sanctions.pm (hunk 5 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm:214]: OK. It will be fixed.

File: lib/Data/Validate/Sanctions/Fetcher.pm (hunk 3 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions/Fetcher.pm:306]: Yes. They are comments or `country (city)` that's better to be chopped off. I'll add a comment.

File: lib/Data/Validate/Sanctions/Fetcher.pm (hunk 1 above)
Comment (felipe-binary) [reply lib/Data/Validate/Sanctions/Fetcher.pm:191]: Such a shame, my eyes deceived me. Sorry about that!

File: lib/Data/Validate/Sanctions.pm (hunk 4 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm]: Fixed.

File: lib/Data/Validate/Sanctions.pm (hunk 5 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm]: Entries are saved in an array with no merging. to simplify lookups, an index is introduced.

File: lib/Data/Validate/Sanctions.pm [hunk 6]
```diff
@@ -73,21 +79,116 @@ sub get_sanction_file {
     return $instance ? $instance->{sanction_file} : $sanction_file;
 }
 
-sub is_sanctioned {        ## no critic (RequireArgUnpacking)
+=head2 get_sanctioned_info
```
Comment (chylli-binary) [inline lib/Data/Validate/Sanctions.pm:82]: function name is wrong here ? 
```suggestion
=head2 is_sanctioned
```

File: lib/Data/Validate/Sanctions.pm [hunk 7]
```diff
@@ -73,21 +79,116 @@ sub get_sanction_file {
...
+    my $args = {};
+
+    # in the new interface we accept fields in a hashref
+    if (ref $_[0] eq 'HASH') {
+        ($args) = @_;
+        ($first_name, $last_name, $date_of_birth) = $args->@{qw/first_name last_name date_of_birth/};
+    }
+
+    # convert country names to iso codes
+    for my $field (qw/place_of_birth residence nationality citizen/) {
```
Comment (chylli-binary) [inline lib/Data/Validate/Sanctions.pm:179]: Is `place_of_birth` a country name too ?

File: lib/Data/Validate/Sanctions.pm [hunk 8]
```diff
@@ -73,21 +79,116 @@ sub get_sanction_file {
...
+        my $value = $args->{$field};
+        next unless $value;
+
+        $args->{$field} = Data::Validate::Sanctions::Fetcher::get_country_code($value);
+    }
 
     unless ($self) {
```
Comment (chylli-binary) [inline lib/Data/Validate/Sanctions.pm:186]: Can we move this piece of code to line https://github.com/binary-com/perl-Data-Validate-Sanctions/pull/34/files#diff-064820bc81e8fbbc057047923c9ea22b512eeb4c4970f2363d384bf946d19746R166 ?

File: t/data/sample_hmt.csv [hunk 9]
```diff
@@ -22,3 +22,4 @@ SO,Sang-kuk,,,,,,00/00/1935,,,,,,"Head of the Department of Nuclear Physics, Kim
...
 PLOTNITSKII,Igor,Venediktovich,,,,,26/06/1964,"Luhansk, possibly in Kelmentsi, Chernivtsi Oblast",Ukraine,,,,,,,,,,,,,Former so-called 'Defence Minister' and former so-called 'Head of the Luhansk People's Republic'. Former Special Envoy of the so-called 'Luhansk People�s Republic' on Minsk implementation.,Individual,AKA,Ukraine (Sovereignty),12/07/2014,18/03/2020,13017
+SAEED,Hafez,Mohammad,,,,,05/06/1950,"Sargodha, Punjab",Pakistan,Pakistani,,3520025509842-7 (Pakistani),Leader of Lashkar-e-Tayyiba (QDe.118),House No 116E,Mohalla Johar,Lahore,Tehsil,Lahore City,Lahore District,123321,Pakistan (location as at May 2008),(UK Sanctions List Ref):AQD0183 (UN Ref):QDi.263. Also referred to as Hafiz Ji.  (Further Identifiying Information):Muhammad Saeed is the leader of Lashkar-e-Tayyiba (QDe.118).INTERPOL-UN Security Council Special Notice web link: https://www.interpol.int/en/notice/search/un/1543488,Individual,AKA,ISIL (Da'esh) and Al-Qaida,12/12/2008,31/12/2020,9215
```
Comment (chylli-binary) [inline t/data/sample_hmt.csv:25]: Add a new line at EOF ?

File: lib/Data/Validate/Sanctions.pm (hunk 7 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm:179]: Yes it is.

Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm:179]: Yes it is.

File: lib/Data/Validate/Sanctions.pm [hunk 10]
```diff
@@ -73,18 +76,114 @@ sub get_sanction_file {
...
+
+It returns 1 if a match is found, otherwise 0.
+
+=cut
+
+sub is_sanctioned {    ## no critic (RequireArgUnpacking)
     return (get_sanctioned_info(@_))->{matched};
 }
 
+sub _match_optional_args {
```
[the remark this answers is not in the document]
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm]: Added.

File: t/data/sample_hmt.csv (hunk 9 above)
Comment (mat-fs) [reply t/data/sample_hmt.csv]: Done.

File: lib/Data/Validate/Sanctions.pm (hunk 8 above)
Comment (mat-fs) [reply lib/Data/Validate/Sanctions.pm]: Moved.

File: bin/update_sanctions_csv [hunk 11]
```diff
@@ -10,12 +10,19 @@ my $help;
 my $eu_token;
```
Comment (billmarriott-binary) [inline bin/update_sanctions_csv:10]: Aren't we going to stop using this file in your other card? https://redmine.deriv.cloud/issues/22618#bug/sanctions_update_cron_failing_silently

File: bin/update_sanctions_csv [hunk 12]
```diff
@@ -10,12 +10,19 @@ my $help;
...
 my $eu_url;
 my $sanction_file;
+my $hmt_url;
+my $ofac_sdn_url;
+my $ofac_consolidated_url;
```
Comment (billmarriott-binary) [inline bin/update_sanctions_csv:15]: I thought this card was just adding more filtering criteria, why are we adding new sources?

File: bin/update_sanctions_csv (hunk 11 above)
Comment (mat-fs) [reply bin/update_sanctions_csv:10]: We are not using it, but it's not a good idea to remove it either IMO.

File: bin/update_sanctions_csv (hunk 12 above)
Comment (mat-fs) [reply bin/update_sanctions_csv:15]: These sources were there from the beginning. I'm just adding new  optional args in order to accelerate `xt/20_update.t` test script (the slow HM T source specifically).

Comment (mat-fs) [reply bin/update_sanctions_csv:15]: But they can be used for other aims.

**Pull request merged**